fennel.lua

```diff
@@ -1124,16 +1125,15 @@ local function compile1(ast, scope, parent, opts)
     local exprs = {}
 
     -- Compile the form
-    if isList(ast) then
-        -- Function call or special form
+    if isList(ast) then -- Function call or special form
+        local len, oldAst = #ast, ast
         assertCompile(#ast > 0, "expected a function, macro, or special to call", ast)
         ast = macroexpand(ast, scope)
-        local len = ast and #ast or 0
+        -- if it's a macro, recur with the expanded ast
+        if (ast ~= oldAst) then return compile1(ast, scope, parent, opts) end
```

Comment:
Shouldn't this conditional be specifically about whether `ast` is a list or not? If the new AST is different from the old AST, but it's still a list, then the recursion is unnecessary, right?

Reply:
Good question - this was sort of my way of cutting the Gordian Knot ;-)

Looking at `compile1` as a whole, this check is occurring inside the `if isList(ast)` check at the very beginning. The macroexpansion is the only way for `ast` to be a non-table in this part of the code, at which point we'd basically want to start fresh branching on what kind of ast node it is. The cleanest way to do this struck mas a just invoking `compile1` again rather than having to restructure the `if/else` branches.